### PR TITLE
Eth reserve address

### PR DESCRIPTION
**Pull request by barakman**

### CODE REVIEW

File: solidity/contracts/BancorNetwork.sol [hunk 1]
```diff
@@ -37,6 +37,7 @@ contract BancorNetwork is IBancorNetwork, TokenHolder, ContractRegistryClient, F
 
     uint256 private constant CONVERSION_FEE_RESOLUTION = 1000000;
     uint256 private constant AFFILIATE_FEE_RESOLUTION = 1000000;
+    IEtherToken private constant ETH_RESERVE = IEtherToken(-1);
```
Comment (yudilevi) [inline solidity/contracts/BancorNetwork.sol:40]: maybe use something more specific like `ETH_RESERVE_ADDRESS` or `ETH_ADDRESS`?
also address should be 0xeeee.... and not 0xffff....

File: solidity/contracts/BancorNetwork.sol [hunk 2]
```diff
@@ -430,7 +431,7 @@ contract BancorNetwork is IBancorNetwork, TokenHolder, ContractRegistryClient, F
         }
         // EtherToken
         else if (etherTokens[_sourceToken]) {
-            // claim the tokens - if the source token is address(0), this call will fail
+            // claim the tokens - if the source token is ETH_RESERVE, this call will fail
```
Comment (yudilevi) [inline solidity/contracts/BancorNetwork.sol:434]: should be more like `if the source token address is ETH_RESERVE_ADDRESS....`

File: solidity/contracts/BancorNetwork.sol [hunk 3]
```diff
@@ -533,9 +534,9 @@ contract BancorNetwork is IBancorNetwork, TokenHolder, ContractRegistryClient, F
         // source is ETH
         ConversionStep memory stepData = data[0];
         if (etherTokens[stepData.sourceToken]) {
-            // newer converter - replace the source token with address(0)
+            // newer converter - replace the source token with ETH_RESERVE
```
Comment (yudilevi) [inline solidity/contracts/BancorNetwork.sol:537]: same as above

File: solidity/contracts/BancorNetwork.sol [hunk 4]
```diff
@@ -544,9 +545,9 @@ contract BancorNetwork is IBancorNetwork, TokenHolder, ContractRegistryClient, F
         // target is ETH
         stepData = data[data.length - 1];
         if (etherTokens[stepData.targetToken]) {
-            // newer converter - replace the source token with address(0)
+            // newer converter - replace the source token with ETH_RESERVE
```
Comment (yudilevi) [inline solidity/contracts/BancorNetwork.sol:548]: same as above

File: solidity/contracts/converter/BancorConverter.sol [hunk 5]
```diff
@@ -27,6 +27,7 @@ contract BancorConverter is IBancorConverter, TokenHandler, SmartTokenController
 
     uint32 private constant WEIGHT_RESOLUTION = 1000000;
     uint64 private constant CONVERSION_FEE_RESOLUTION = 1000000;
+    IERC20Token private constant ETH_RESERVE = IERC20Token(-1);
```
Comment (yudilevi) [inline solidity/contracts/converter/BancorConverter.sol:30]: see comment in the network

File: solidity/contracts/helpers/TestBancorConverter.sol [hunk 6]
```diff
@@ -12,7 +12,6 @@ contract TestBancorConverter is BancorConverter {
         BancorConverter(_token, _registry, _maxConversionFee, _reserveToken, _reserveWeight)
         public
     {
-        etherToken = IEtherToken(0);
```
Comment (yudilevi) [inline solidity/contracts/helpers/TestBancorConverter.sol:15]: don't we still need this?

File: solidity/test/BancorConverter.js [hunk 7]
```diff
@@ -55,8 +55,8 @@ async function initConverter(accounts, activate, maxConversionFee = 0) {
     return converter;
 }
 
-function verifyReserve(reserve, virtualBalance, weight, isSet) {
-    assert.equal(reserve[0], virtualBalance);
+function verifyReserve(reserve, balance, weight, isSet) {
```
Comment (yudilevi) [inline solidity/test/BancorConverter.js:58]: do we actually need to test the balance here? isn't it deprecated in the reserve struct?

File: solidity/test/BancorConverterUpgrader.js [hunk 8]
```diff
@@ -14,6 +14,8 @@ const ContractFeatures = artifacts.require('ContractFeatures');
 const BancorConverterFactory = artifacts.require('BancorConverterFactory');
 const BancorConverterUpgrader = artifacts.require('BancorConverterUpgrader');
 
+const ETH_RESERVE = '0x'.padEnd(42, 'f');
```
Comment (yudilevi) [inline solidity/test/BancorConverterUpgrader.js:17]: address should be 0xeeee....

File: solidity/test/BancorConverterWithEthReserve.js [hunk 9]
```diff
@@ -17,6 +17,8 @@ const EtherToken = artifacts.require('EtherToken');
 const BancorConverterFactory = artifacts.require('BancorConverterFactory');
 const BancorConverterUpgrader = artifacts.require('BancorConverterUpgrader');
 
+const ETH_RESERVE = '0x'.padEnd(42, 'f');
```
Comment (yudilevi) [inline solidity/test/BancorConverterWithEthReserve.js:20]: should be 0xeeee....

File: solidity/test/BancorNetwork.js [hunk 10]
```diff
@@ -450,13 +452,13 @@ contract('BancorNetwork', accounts => {
...
         it('should throw when calling convertFor with invalid path', async () => {
-            let invalidPath = [utils.zeroAddress, smartToken1.address];
+            let invalidPath = [ETH_RESERVE, smartToken1.address];
             await utils.catchRevert(bancorNetwork.convertFor(invalidPath, 10000, 1, accounts[1], { value: 10000 }));
         });
 
         it('should throw when calling convertFor with invalid long path', async () => {
             let longBuyPath = [];
-            for (let i = 0; i < 100; ++i)
+            for (let i = 0; i < 101; ++i)
```
Comment (yudilevi) [inline solidity/test/BancorNetwork.js:461]: why change from 100 to 101?

File: solidity/contracts/helpers/TestBancorConverter.sol (hunk 6 above)
Comment (barakman) [reply solidity/contracts/helpers/TestBancorConverter.sol:15]: @yudilevi : I though about it for a while. Definitely not `0` (if anything, the real address, i.e., 0xeee...). But it's not really used anyway (i.e., every test which uses this helper contract, calls the setter function immediately after instantiating the contract). I'll do another sweep in order to make sure of that.

File: solidity/test/BancorConverter.js (hunk 7 above)
Comment (barakman) [reply solidity/test/BancorConverter.js:58]: @yudilevi : No, it actually isn't. We've replaced the (not quite used) virtual balance with the (now very much in use) real balance.

File: solidity/test/BancorNetwork.js (hunk 10 above)
Comment (barakman) [reply solidity/test/BancorNetwork.js]: @yudilevi : Did some cleanup around this; please see updated code.

File: solidity/contracts/helpers/TestBancorConverter.sol (hunk 6 above)
Comment (yudilevi) [reply solidity/contracts/helpers/TestBancorConverter.sol:15]: ok cool

File: solidity/test/BancorConverter.js (hunk 7 above)
Comment (yudilevi) [reply solidity/test/BancorConverter.js:58]: ahh right

File: solidity/test/BancorNetwork.js (hunk 10 above)
Comment (yudilevi) [reply solidity/test/BancorNetwork.js]: ok

File: solidity/contracts/BancorNetwork.sol (hunk 1 above)
Comment (barakman) [reply solidity/contracts/BancorNetwork.sol]: @yudilevi : Please see updated code.

File: solidity/contracts/BancorNetwork.sol (hunk 2 above)
Comment (barakman) [reply solidity/contracts/BancorNetwork.sol]: @yudilevi : Please see updated code.

File: solidity/contracts/BancorNetwork.sol (hunk 3 above)
Comment (barakman) [reply solidity/contracts/BancorNetwork.sol]: @yudilevi : Please see updated code.

File: solidity/contracts/BancorNetwork.sol (hunk 4 above)
Comment (barakman) [reply solidity/contracts/BancorNetwork.sol]: @yudilevi : Please see updated code.

**Pull request merged**